politeiad/backendv2/tstorebe/plugins/ticketvote/cmds.go

```diff
@@ -0,0 +1,2976 @@
+	// out.
+	lf, err := p.submissionsCache(token)
+	if err != nil {
+		return nil, err
+	}
+	expected := make(map[string]struct{}, len(lf.Tokens)) // [token]struct{}
+	for k := range lf.Tokens {
+		token, err := tokenDecode(k)
+		if err != nil {
+			return nil, err
+		}
+		r, err := p.recordAbridged(token)
+		if err != nil {
+			return nil, err
+		}
+		if r.RecordMetadata.Status != backend.StatusPublic {
+			// This record is not public and should not be included
+			// in the runoff vote.
+			continue
+		}
+
+		// This is a public record that is part of the parent record's
+		// submissions list. It is required to be in the runoff vote.
+		expected[k] = struct{}{}
+	}
+
+	// Verify that there are no extra submissions in the runoff vote
+	for _, v := range s.Starts {
+		_, ok := expected[v.Params.Token]
+		if !ok {
+			// This submission should not be here
+			return nil, backend.PluginError{
+				PluginID:  ticketvote.PluginID,
+				ErrorCode: uint32(ticketvote.ErrorCodeStartDetailsInvalid),
+				ErrorContext: fmt.Sprintf("record %v should not be included",
+					v.Params.Token),
+			}
+		}
+	}
+
+	// Verify that the runoff vote is not missing any submissions
+	subs := make(map[string]struct{}, len(s.Starts))
+	for _, v := range s.Starts {
+		subs[v.Params.Token] = struct{}{}
+	}
+	for k := range expected {
+		_, ok := subs[k]
+		if !ok {
+			// This records is missing from the runoff vote
+			return nil, backend.PluginError{
+				PluginID:     ticketvote.PluginID,
+				ErrorCode:    uint32(ticketvote.ErrorCodeStartDetailsMissing),
+				ErrorContext: k,
+			}
+		}
+	}
+
+	// Prepare start runoff record
+	submissions := make([]string, 0, len(subs))
+	for k := range subs {
+		submissions = append(submissions, k)
+	}
+	srr = &startRunoffRecord{
+		Submissions:      submissions,
+		Mask:             mask,
+		Duration:         duration,
+		QuorumPercentage: quorum,
+		PassPercentage:   pass,
+		StartBlockHeight: sr.StartBlockHeight,
+		StartBlockHash:   sr.StartBlockHash,
+		EndBlockHeight:   sr.EndBlockHeight,
+		EligibleTickets:  sr.EligibleTickets,
+	}
+
+	// Save start runoff record
+	err = p.startRunoffRecordSave(treeID, *srr)
+	if err != nil {
+		return nil, fmt.Errorf("startRunoffRecordSave %v: %v",
+			treeID, err)
+	}
+
+	return srr, nil
+}
+
+// startRunoff starts a runoff vote.
+func (p *ticketVotePlugin) startRunoff(treeID int64, token []byte, s ticketvote.Start) (*ticketvote.StartReply, error) {
+	// Sanity check
+	if len(s.Starts) == 0 {
+		return nil, fmt.Errorf("no start details found")
+	}
+
+	// Perform validation that can be done without fetching any records
+	// from the backend.
+	var (
+		mask     = s.Starts[0].Params.Mask
+		duration = s.Starts[0].Params.Duration
+		quorum   = s.Starts[0].Params.QuorumPercentage
+		pass     = s.Starts[0].Params.PassPercentage
+		parent   = s.Starts[0].Params.Parent
+	)
+	for _, v := range s.Starts {
+		// Verify vote params are the same for all submissions
+		switch {
```

Comment:
This looks like duplicate code as well.

Reply:
Its not.  This validation only applies to runoff votes and only appears once.